backend/src/main/java/org/cryptomator/hub/api/UsersResource.java

```diff
@@ -71,14 +70,33 @@ public Response putMe(@Nullable @Valid UserDto dto) {
 		user.setPictureUrl(jwt.getClaim("picture"));
 		user.setEmail(jwt.getClaim("email"));
 		if (dto != null) {
-			user.setPublicKey(dto.publicKey);
-			user.setPrivateKey(dto.privateKey);
+			user.setEcdhPublicKey(dto.ecdhPublicKey);
+			user.setEcdsaPublicKey(dto.ecdsaPublicKey);
+			user.setPrivateKeys(dto.privateKeys);
 			user.setSetupCode(dto.setupCode);
+			updateDevices(user, dto);
 		}
 		userRepo.persist(user);
 		return Response.created(URI.create(".")).build();
 	}
 
+	/**
+	 * updates those devices that are present in both the entity and the dto. No devices are added or removed.
+	 * @param userEntity The persistent entity
+	 * @param userDto The DTO
+	 */
+	private void updateDevices(User userEntity, UserDto userDto) {
+		var dtos = userDto.devices.stream().collect(Collectors.toMap(DeviceResource.DeviceDto::id, Function.identity()));
+		var updatedDevices = userEntity.devices.stream().filter(d -> dtos.containsKey(d.getId())).peek(device -> {
+			var dto = dtos.get(device.getId());
+			device.setType(dto.type());
+			device.setName(dto.name());
+			device.setPublickey(dto.publicKey());
+			device.setUserPrivateKeys(dto.userPrivateKeys());
+		});
+		deviceRepo.persist(updatedDevices);
+	}
```

Comment:
I agree that `peek(...)` is not optimal. I considered using `map(device -> { ... return device; })` instead.

However `foreach` is not good either, as it is a terminal operation and requires multiple calls to `deviceRepo.persist(...)`, while the implemented solution can bulk-persist.

Reply:
A different static code analysis tool isn't happy about using `map` when it returns its parameter. However, how about inverting the logic to find matching pairs of DTOs and Device entities:

In line 89, let's create a lookup table from `userEntity.devices` instead of `userDto.devices`.

Then in the following lines start by streaming `userDto.devices.stream().filter(d -> devices.containsKey(d.id()))` and map from dto to device. This should make the other tool happy as well.

Reply:
changed in 6da692e